### PR TITLE
WIP: Feature/ssh support

### LINKED ISSUE #283: Support own SSH keys at cluster creation
We need to support the saved SSH keys.
* The user needs to add a key to a cluster
* When the key is deleted all other clusters using this key need to drop the refernece.

### PR DESCRIPTION
**Pull request by lbb**
Closes #283

### CODE REVIEW

File: handler/cluster.go [hunk 1]
```diff
@@ -17,6 +18,7 @@ import (
 func newClusterEndpointV2(
 	kps map[string]provider.KubernetesProvider,
 	dcs map[string]provider.DatacenterMeta,
+	clientset extensions.Clientset,
```
Comment (mrIncompetent) [inline handler/cluster.go]: Name is ambiguous

File: handler/cluster.go [hunk 2]
```diff
@@ -73,6 +76,26 @@ func newClusterEndpointV2(
...
+		// TODO(realfake): Duplicated code move to function
+		sshClient := clientset.SSHKeyTPR(req.user.Name)
+		keys, err := sshClient.List()
+		if err != nil {
+			return nil, err
+		}
+
+		for _, key := range keys.Items {
+			for _, sshKeyName := range req.SSHKeys {
+				if sshKeyName == key.Metadata.Name {
```
Comment (mrIncompetent) [inline handler/cluster.go]: https://blog.codinghorror.com/flattening-arrow-code/

"guard clauses"

File: handler/cluster.go [hunk 3]
```diff
@@ -244,6 +268,26 @@ func deleteClusterEndpoint(
...
+		// TODO(realfake): Duplicated code move to function
+		sshClient := clientset.SSHKeyTPR(req.user.Name)
+		keys, err := sshClient.List()
+		if err != nil {
+			return nil, err
+		}
+
+		for _, key := range keys.Items {
+			for i, clusterName := range key.Clusters {
+				if clusterName == req.cluster {
```
Comment (mrIncompetent) [inline handler/cluster.go]: https://blog.codinghorror.com/flattening-arrow-code/

"guard clauses"

File: handler/cluster_test.go [hunk 4]
```diff
@@ -1,17 +1,6 @@
...
-	"net/http/httptest"
-	"testing"
-
-	"bytes"
-
-	"github.com/kubermatic/api"
-)
-
+/*
 func TestNewClusterEndpoint(t *testing.T) {
```
Comment (mrIncompetent) [inline handler/cluster_test.go]: Why was this commented out?
Avoid committing commented out code.

File: handler/nodes_test.go [hunk 5]
```diff
@@ -1,15 +1,6 @@
...
-	"fmt"
-	"net/http"
-	"net/http/httptest"
-	"testing"
-
-	"github.com/kubermatic/api"
-)
-
+/*
 func TestCreateNodesEndpoint(t *testing.T) {
```
Comment (mrIncompetent) [inline handler/nodes_test.go]: Why was this commented out?
Avoid committing commented out code.

File: provider/cloud/aws/provider.go [hunk 6]
```diff
@@ -543,6 +573,20 @@ func (a *aws) CreateNodes(ctx context.Context, cluster *api.Cluster, node *api.N
 	if err != nil {
 		return nil, fmt.Errorf("failed get ec2 client: %v", err)
 	}
+	var keyname string
+	if len(keys) < 1 {
+		glog.Info("could not find keypair, fallback")
```
Comment (mrIncompetent) [inline provider/cloud/aws/provider.go]: Why couldn't i find a keypar? What fallback?

File: provider/cloud/aws/provider.go [hunk 7]
```diff
@@ -543,6 +573,20 @@ func (a *aws) CreateNodes(ctx context.Context, cluster *api.Cluster, node *api.N
...
+		keyname = cluster.Spec.Cloud.AWS.SSHKeyName
+	} else {
+		keyname, _, err = getOrCreateKey(keys, client)
+	}
+	if err != nil {
```
Comment (mrIncompetent) [inline provider/cloud/aws/provider.go]: What is this for?

File: provider/cloud/aws/provider.go [hunk 8]
```diff
@@ -531,7 +533,35 @@ func (a *aws) GetContainerLinuxAmiID(version string, client *ec2.EC2) (string, e
 	return *latestImage.ImageId, nil
 }
 
-func (a *aws) CreateNodes(ctx context.Context, cluster *api.Cluster, node *api.NodeSpec, num int) ([]*api.Node, error) {
+func getOrCreateKey(keys []extensions.UserSSHKey, client *ec2.EC2) (name string, created bool, err error) {
```
Comment (mrIncompetent) [inline provider/cloud/aws/provider.go:563]: If you already define named return values, use them.

File: provider/cloud/digitalocean/provider.go [hunk 9]
```diff
@@ -163,12 +168,29 @@ func (do *digitalocean) CreateNodes(ctx context.Context, cluster *api.Cluster, s
 		t := token(cSpec.GetToken())
 		client := godo.NewClient(oauth2.NewClient(ctx, t))
 
+		var dropKeys []string
+		doKeys, _, err := client.Keys.List(&godo.ListOptions{
+			Page:    1,
+			PerPage: 50,
```
Comment (mrIncompetent) [inline provider/cloud/digitalocean/provider.go]: Why do you request a paginated response but don't handle the case if you get more than 1 page?
Either don't request a paginated response (preferred in this case)
Or handle multiple pages

**Review by mrIncompetent — CHANGES_REQUESTED** [review]

File: handler/cluster.go (hunk 2 above)
Comment (lbb) [reply handler/cluster.go]: This is not possible, thus it is a loop and not a nested if. I could write it into a own function.
The loops are used to filter elements.

Comment (lbb) [reply handler/cluster.go]: This is not possible, thus it is a loop and not a nested if. I could write it into a own function.
The loops are used to filter elements.

File: handler/cluster.go (hunk 3 above)
Comment (lbb) [reply handler/cluster.go]: This is not possible, thus it is a loop and not a nested if. I could write it into a own function.
The loops are used to filter elements.

File: handler/cluster_test.go (hunk 4 above)
Comment (lbb) [reply handler/cluster_test.go]: I need to write a fake TPRClientset, otherwise the tests will fail

File: handler/nodes_test.go (hunk 5 above)
Comment (lbb) [reply handler/nodes_test.go]: See above

File: provider/cloud/aws/provider.go (hunk 8 above)
Comment (lbb) [reply provider/cloud/aws/provider.go:563]: Please recheck. I'd rather use full return statements for expressiveness and readability. Especially in a big function

File: handler/cluster.go (hunk 2 above)
Comment (mrIncompetent) [reply handler/cluster.go]:
```golang
if sshKeyName != key.Metadata.Name {
    continue
}
key.Clusters = append(key.Clusters, c.Metadata.Name)
// TODO(realfake): This takes a long time look forward to async / batch implementation
_, err := sshClient.Update(&key)
if err != nil {
    return nil, err
}
```

File: handler/cluster.go (hunk 3 above)
Comment (mrIncompetent) [reply handler/cluster.go]: See https://github.com/kubermatic/api/pull/312/files#r110847951

File: handler/cluster_test.go (hunk 4 above)
Comment (mrIncompetent) [reply handler/cluster_test.go]: So, you are not done with this PR?

File: handler/nodes_test.go (hunk 5 above)
Comment (mrIncompetent) [reply handler/nodes_test.go]: So, you are not done with this PR?

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**